fix functional collective to not generate getattr node

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97945
* __->__ #97924

use mesh.get_dim_groups directly instead of doing mesh tensor operations

This help us get rid of the getattr ops during tracing

wanchaol committed:
fix functional collective to not generate getattr node

use mesh.get_dim_groups directly instead of doing mesh tensor operations

This help us get rid of the getattr ops during tracing

[ghstack-poisoned]

wanchaol committed:
Update on "fix functional collective to not generate getattr node"


use mesh.get_dim_groups directly instead of doing mesh tensor operations

This help us get rid of the getattr ops during tracing

[ghstack-poisoned]

wanchaol committed:
Update on "fix functional collective to not generate getattr node"


use mesh.get_dim_groups directly instead of doing mesh tensor operations

This help us get rid of the getattr ops during tracing

[ghstack-poisoned]

wanchaol committed:
Update on "fix functional collective to not generate getattr node"


use mesh.get_dim_groups directly instead of doing mesh tensor operations

This help us get rid of the getattr ops during tracing

[ghstack-poisoned]

kumpera commented on torch/distributed/_functional_collectives.py:
```diff
@@ -226,17 +226,20 @@ def _expand_group(group: RANK_TYPES, tag: str = "") -> Tuple[str, List[int], int
-            rankset = dmesh.mesh.swapdims(-1, dim).reshape(-1, group_size).flatten().tolist()
-            tag = tag or c10d._get_group_tag(dmesh.get_dim_groups()[dim])
+            dim_group = dmesh.get_dim_groups()[dim]
+            rankset = dist.get_process_group_ranks(dim_group)
```
This prevents us from later doing collective algebraic fusion as the compiler won't see the whole mesh.

wanchaol replied:
Yeah we won't see the whole mesh, but given that this branch is taking a tuple of mesh, mesh_dim, it should be the correct behavior. 


filed an separate issue to track the whole mesh https://github.com/pytorch/pytorch/issues/97991.

kumpera reviewed (approved):
This PR removes some capabilities we current have and I rather have us fix this in another way. If that's not possible, we should at least file an issue so we don't forget what we're losing here.

wanchaol committed:
Update on "fix functional collective to not generate getattr node"


use mesh.get_dim_groups directly instead of doing mesh tensor operations

This help us get rid of the getattr ops during tracing

[ghstack-poisoned]

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -226,17 +226,20 @@ def _expand_group(group: RANK_TYPES, tag: str = "") -> Tuple[str, List[int], int
-        rankset = group.mesh.swapdims(-1, 0).reshape(-1, group_size).flatten().tolist()
-        tag = tag or c10d._get_group_tag(group.get_dim_groups()[0])
+        assert group.ndim == 1, "Only 1D mesh is supported, pass in (DeviceMesh, int) together if mesh > 1D"
+        # TODO: it should run collective in the whole mesh instead of dim 0
```
I think tracer would just brutally trace what ever the function passed in, in this case `mesh.collective`  -> `funccol.collective`, so it would still trace this function as it, that's why we were generating all those mesh tensor CPU manipulation getattr/constants. I don't know if there's a way for `make_fx` to ignore certain piece of code for the function, so we should probably keep it tracing friendly

wanchaol commented:
@pytorchbot merge -f "windows failure not related"
